ED/protocol.html

```diff
@@ -746,7 +746,7 @@ <h3 property="schema:name">HTTP Server</h3>
 
                   <p about="" id="server-content-type-includes" rel="spec:requirement" resource="#server-content-type-includes"><span property="spec:statement"><span rel="spec:requirementSubject" resource="#Server">Server</span> <span rel="spec:requirementLevel" resource="spec:MUST">MUST</span> generate a <code>Content-Type</code> header field in a message that contains content.</span></p>
 
-                  <p about="" id="server-content-type-missing" rel="spec:requirement" resource="#server-content-type-missing"><span property="spec:statement"><span rel="spec:requirementSubject" resource="#Server">Server</span> <span rel="spec:requirementLevel" resource="spec:MUST">MUST</span> reject <code>PUT</code>, <code>POST</code> and <code>PATCH</code> requests without the <code>Content-Type</code> header with a status code of <code>400</code>.</span> [<a href="https://github.com/solid/specification/issues/70#issuecomment-547924171" rel="cito:citesAsSourceDocument">Source</a>]</p>
+                  <p about="" id="server-content-type-missing" rel="spec:requirement" resource="#server-content-type-missing"><span property="spec:statement"><span rel="spec:requirementSubject" resource="#Server">Server</span> <span rel="spec:requirementLevel" resource="spec:MUST">MUST</span> reject <code>PUT</code>, <code>POST</code> and <code>PATCH</code> requests including content but without the <code>Content-Type</code> header with a status code of <code>400</code>.</span> [<a href="https://github.com/solid/specification/issues/70#issuecomment-547924171" rel="cito:citesAsSourceDocument">Source</a>]</p>
```

Comment:
Suggest making the language more RFC9110 like and also include validity as a condition:

"On receiving a message containing content but not including a content-type header field with a valid media type in a PUT, POST or PATCH request, the server MUST reject the request with a status code of 400."

Reply:
That seems like proposing a new requirement, "I guess we'll need a separate PR for that!"

But here is why I think it'd be preferable to not introduce that:

RFC doesn't explicitly require recipients to verify the content against the `Content-Type` header field value. Some servers may do that for practical reasons, if performance is not an issue, or if precision is expected by the server before storing the state, and so on. By definition, the `Content-Type` header field indicates the intended media type. It doesn't need to be further spelled out beyond what the RFC says.

We are highly unlikely to have a complete test coverage taking input payload and media type and checking whether they match, and this gets potentially more complex with polyglot formats.

Validity of the `Content-Type` header field value as a condition would in fact be a separate requirement than what's originally intended by the #server-content-type-missing requirement, and it would not entail `400` when the request is rejected either.

In cases that "the origin server is refusing to service the request because the content is in a format not supported by this method on the target resource", the 415 status code can be used.

So, I do not think that the requirement needs to validating the `Content-Type` header field-value against the content. It is already implicitly covered.

---

I don't think this (or any) requirement should be more RFC9110, not that there is anything wrong with that. The language in this document is aimed to be internally consistent first and foremost, and the conformance clause template that's in use is along the lines what we see right now. There are variations of course throughout the spec.

I have changed "including" to "containing" and added "field" to the text as per your suggestion. Thanks.

Reply:
I think you misunderstood my worry here, I should have explained this better.

RFC9110, 8.3 in the paragraph quoted, is a requirement for the clients to provide a content-type field. Translating this for a server, not only should the content-type field be present, but it also must be in the correct syntax. 

It is NOT my intention to impose verification requirements on the server that the content-type matches the content. This is a purely header field syntax check!

Reply:
Fair enough, but I think that level of correctness is already expected of Solid servers: https://solidproject.org/ED/protocol#http-server . And even more so that I've raised it as a significant security consideration with some advisement: https://solidproject.org/ED/protocol#http-request-headers-validity . 400 is indeed expected in these cases.

Reply:
I think there is value to making this explicit to spec readers here. Rather than have them co-relate this in conjunction with security consideration (which is non-normative) and say, leading to a scenario where even one implementor amongst many will tolerate not well-formed content-type header, which is certainly not ideal. It also demands that the test suite include this case as an explicit test.

Thinking about it, I would ideally want servers to be strict not just on syntax but also on the iana registered types (but not subtypes) when processing the content type headers. Not sure how to work that in or even if that is advisable?

Reply:
Correct use of HTTP is already required for interoperability:

* https://solidproject.org/ED/protocol#server-http
* https://solidproject.org/ED/protocol#client-http

Full stop.

*Advisements* under (security) considerations does not take away from *requirements*.

Whether an implementation of the Solid https://solidproject.org/ED/qa#TestSuite has its own https://solidproject.org/ED/qa#TestCase s checking HTTP conformance or imports an existing one is orthogonal to this matter. A conforming Solid Protocol product means that the outcome of applicable and testable tests is that the given product meets all the requirements.

Reply:
Sorry, but a RFC9110 compliant server need not send a 400 error code when sent a request with a bad/illegal `content-type` header, it can ignore it and send a 2xx response. And since Solid server extends an HTTP server, a Solid server can ipso-facto ignore it too (not withstanding the MUST for specifying a `content-type` header, which merely says that the header field has to be there not that field value has to be legal). Hence, this needs to be explicitly constrained.

This is not about just correct use of HTTP, but application of a rule that is more stringent in Solid than HTTP, which must be thoroughly specified.

Reply:
I see no practical or theoretical benefit to further constraining the media type in the Solid Protocol. Just because a client includes a "legal" media type - whatever that is - it does not entail that a server accepts it. The server already has the liberty to accept or reject a request for various reasons and it is useful to maintain that right. Some servers may only acknowledge IANA registered media types, while others may allow other media types, yet some other servers may not want to handle multimedia formats, and some may be specialised to to super cool things with them, and so on. 415 exists. This is always a negotiation between the server-client pair, and it does not conflict with the kind of interoperability we can expect from the Solid Protocol.

Reply:
> Just because a client includes a "legal" media type - whatever that is...

Do not strawman the argument. I am speaking of an illegal `content-type` header, not `illegal` media-type. As I have said, "_The suggested is a purely header field syntax check_". 

For IANA registered types (not even subtypes), I already indicated that it is a preference, which I am not sure how that would even work.  This was a separate musing (in a separate paragraph) from the request of making sure that the `content-type` header is syntactically correct.

> The server already has the liberty to accept or reject a request for various reasons and it is useful to maintain that right.

So you want to maintain the right of some random Solid server to accept a request with an illegal `content-type` header despite the change in Solid protocol. I cannot think of one situation where that is ever useful.

Yes, I want to constrain the server behaviour because it reflects the constraint added to client behaviour by Solid protocol, a constraint that HTTP itself does not have.

Reply:
Excuse me, but I've been trying very hard to understand your concerns, and it still doesn't make sense to me.

You're using terminology like "valid media type," "correct syntax," "head field syntax check," "well-formed content-type header," "field value has to be legal", "syntactically correct." However, you also seem to find the HTTP definitions, such as the ABNF notations and conformance explanations, inadequate for defining what constitutes "valid," "correct," "legal," or similar terms for both the header field and its field value. You've been jumping back and forth, in addition to suggesting that the requirement level is relevant. When I tried to understand your statement "field value has to be legal" - again, whatever "legal" means (perhaps you should define it?) - with the possibility that there is a finite set of media types or categories of them, you accused me of straw-manning.

How about them straw-manning:

>So you want to maintain the right of some random Solid server to accept a request with an illegal content-type header despite the change in Solid protocol. I cannot think of one situation where that is ever useful.

How did you conclude that I want the server to accept an "illegal content-type header" when I literally argued earlier that there is nothing for the Solid Protocol to do here since it is building on HTTP by definition, and that products either conform to HTTP or not?

https://www.rfc-editor.org/rfc/rfc9110#section-5.1-3

https://www.rfc-editor.org/rfc/rfc9110#section-8.3-2

```
Content-Type = media-type
```

follow the white-rabbit for conformance.

So, at this juncture, I'm not sure if there is any new information I can add to this thread in this particular PR. Perhaps we need more bandwidth to discuss or draw things on a whiteboard. If you feel this is inadequate, please create a new issue and make your case there, kick off an STM, etc.

Reply:
Our disagreement here boils down to a very simple interpretation of how a change in client (sender for PUT, POST, PATCH) requirement translates into a change in server requirements. Where you reverse the Strawman accusation on me is exactly the thing I find to be the problem of interpretation. AFAICT, nothing in plain RFC9110 stops a server from processing a message with a `content-type` type header field that contains anything other than a media type and sending a 2xx response. Due to the product class definition, I would presume that would extend to a Solid server unless explicitly prohibited. You OTOH think that the CoP definition of server somehow also demands that the server sends a 400 in that case, which I do not follow! 

Maybe a higher bandwidth communication will help, I am open to that. But my worry is that, if this is an interpretation of formal language, that bandwidth will not do much to help. 

PS: You may feel that you are trying really hard to understand me and knowing you, I believe that you are, but the kind of "escalation dominance" that you engage in does not help!